### ISSUE TITLE
Support for docker_repository_image copying tagged items

### ISSUE THREAD

**Issue by stephenliberty** [kreuzwerker/terraform-provider-docker#137]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

TL;DR version is that in some commercial cases you have a "trusted repository" (let's say artifactory.foo.com) that your production bits are sent to, scanned, whatever.. and then this gets replicated to other repositories (for instance, AWS ECR). Being able to "download" an image, tag it, and send it elsewhere would be great. For example, verifying that the "latest" in ECR repo 1 is the same as latest in our example artifactory.foo.com. It seems like most of the pieces are there, it's just that there's an expectation that this will run a full build step which in this case is unnecessary. 

### New or Affected Resource(s)

I believe this would probably just be docker_repository_image

### Potential Terraform Configuration

```hcl
# Copy-paste your Terraform configurations here - for large Terraform configs,
# please use a service like Dropbox and share a link to the ZIP file. For
# security, you can also encrypt the files using our GPG public key.

data "docker_registry_image" "upstream_image" {
  name = "artifactory.foo.com/docker/nginx:1.2.3"
}

resource "docker_image" "upstream_image" {
  name          = data.docker_registry_image.upstream_image.name
  pull_triggers = [data.docker_registry_image.upstream_image.sha256_digest]
}

resource "docker_registry_image" "run_image" {
   name = "some.ecr.repository/my-nginx/golden:latest"
   source = docker_image.upstream_image.latest
}

```

### References

* #0000

**Comment by suzuki-shunsuke** [comment]
Please let me confirm what you want.
In the above case, do you want to pull `artifactory.foo.com/docker/nginx:1.2.3` and push it to `some.ecr.repository/my-nginx/golden:latest` without building Docker image?

**Comment by stephenliberty** [comment]
Exactly that, yes. I found I could sort of do that by adding a Dockerfile to the repo referring to that `artifactory.foo.com/docker/nginx:1.2.3` as the base image and adding nothing else, but it feels like an unnecessary hack, ya know?

**Comment by suzuki-shunsuke** [comment]
I see. Thanks.
I think your idea is good as a work around.

I understand your proposal, but I don't know whether this provider should support this feature.
Generally I think we should be careful when we add a new feature for simplicity and maintainability.
If there are other services or tools, maybe we don't have to support this feature.

I think Terraform isn't appropriate to mirror Docker image in real time.

@mavogel Do you have any opinion?

**Comment by ojacobson** [comment]
For what it's worth, I also need a feature like this. I'm working around it right now using a null_resource and an external data provider to re-tag the target image every time the source image changes, and push it, but it feels like a huge hack and leaves a lot of litter in the target repository.

The workflow in my case is that project A builds an `example/runtime:foo` image and publishes it to Docker Hub. This image contains an AWS Lambda function. Project B creates an AWS ECR repository, and re-tags `example/runtime:foo` as `my-account-id.dkr.ecr.some-region.amazonaws.com/a-generated-name:foo` before pushing it.

Using a shell script, this is easy:

```bash
docker pull example/runtime:foo
docker tag example/runtime:foo my-account-id.dkr.ecr.some-region.amazonaws.com/a-generated-name:foo
docker push my-account-id.dkr.ecr.some-region.amazonaws.com/a-generated-name:foo
```

Using the Docker provider, this is … if not impossible, then at least very non-obvious.

I'd prefer not to make Project B start building Docker images, as the image already exists. I only need to add a tag so that I can push the image to the generated ECR repository.

**Comment by acidprime** [comment]
Yeah just to chime in, I have having I think more or less the same difficulty. I think you could also consider just adding a `docker_tag` resource that would do the same as `docker tag example/runtime:foo my-account-id.dkr.ecr.some-region.amazonaws.com/a-generated-name:foo` and that would solve my issue. I am trying to get the new AWS lambda + docker support and without this there is allot of chicken before the egg issues.

**Comment by stephenliberty** [comment]
@acidprime I think we're basically thinking the same thing.

The workaround has been sorta obnoxious - I've got a different part of my deploy job that installs docker, pulls the target image, and then pushes that to the 'next level' repo. But if this were a new setup, I wouldn't be able to do that because the repo had not yet been created by terraform. It's kind of a strange problem to be able to do next to everything in terraform, but not be able to do this cleanly.

(To be clear, I'd love to try to do another MR, but Go is not my forte and I've got zero bandwidth to try and tackle this at the moment)
